src/Plugin/PasswordConstraint/SpecialCharacter.php

```diff
@@ -63,20 +64,47 @@ public function settingsForm(): array {
   /**
    * {@inheritdoc}
    */
-  public function settingsValidate(array &$form, FormStateInterface $form_state): void {
-    parent::settingsValidate($form, $form_state);
+  public function validateConfigurationForm(array &$form, FormStateInterface $form_state): void {
+    parent::validateConfigurationForm($form, $form_state);
 
-    if (empty($form_state->getValue(['settings', 'use_custom_special_characters']))) {
+    if (empty($form_state->getValue('use_custom_special_characters'))) {
       $input = &$form_state->getUserInput();
-      unset($input['settings']['special_characters']);
-      $form_state->setValue(['settings', 'special_characters'], NULL);
-      $form_state->setValue(['settings', 'use_custom_special_characters'], 0);
+      unset($input['special_characters']);
+      $form_state->setValue('special_characters', NULL);
+      $form_state->setValue('use_custom_special_characters', 0);
     }
-    elseif (preg_match('/[a-z0-9]/i', $form_state->getValue(['settings', 'special_characters']))) {
+    elseif (preg_match('/[a-z0-9]/i', $form_state->getValue('special_characters'))) {
       $form_state->setError($form['special_characters'], $this->t('Alphanumeric characters are not allowed.'));
     }
   }
 
+  /**
+   * {@inheritdoc}
+   */
+  public function submitConfigurationForm(array &$form, FormStateInterface $form_state): void {
+    parent::submitConfigurationForm($form, $form_state);
+    if ($form_state->getValue('use_custom_special_characters')) {
+      $this->configuration['use_custom_special_characters'] = TRUE;
+      $this->configuration['special_characters'] = $form_state->getValue('special_characters');
+    }
+    else {
+      $this->configuration['use_custom_special_characters'] = FALSE;
+      unset($this->configuration['special_characters']);
+    }
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getSummary(): array {
+    $summary = parent::getSummary();
+    $summary['#items'][] = $this->t('Define special characters: %characters', ['%characters' => $this->configuration['use_custom_special_characters'] ? $this->t('Yes') : $this->t('No')]);
```

Comment:
Not sure whether we should deal with this Drupal 8 design issue here again, but this `$this->t()` call also secretly calls the static service container to retrieve the string translation service. The previously suggested https://www.previousnext.com.au/blog/safely-extending-drupal-8-plugin-classes-without-fear-of-constructor-changes injection is probably to only way to DI the service here.